core/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -589,10 +600,10 @@ public void process(@Nullable MappingMetaData mappingMd, boolean allowIdGenerati
         }
 
         // generate id if not already provided and id generation is allowed
-        if (allowIdGeneration) {
-            if (id == null) {
-                id(UUIDs.base64UUID());
-            }
+        if (allowIdGeneration && id == null) {
+            assert autoGeneratedTimestamp == -1;
+            autoGeneratedTimestamp = System.currentTimeMillis();
```

Comment:
and with more thought I think I might have totally misunderstood you and you worried about the default of -1 being "not set".  I agree that in that case we should use max(0, time) here.